# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [91]:
# Imports section
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler


## 1. Loading the dataset

In [55]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

df = pd.read_csv('science_data_large.csv')

df.head(15)

,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


In [57]:
df.describe()

,Temperature °C,Mols KCL,Size nm^3
count,1000.000000,1000.000000,1.000000e+03
mean,500.500000,471.530000,5.086111e+05
std,288.819436,288.482872,4.474838e+05
min,1.000000,1.000000,1.611429e+01
25%,250.750000,226.750000,1.298267e+05
50%,500.500000,459.500000,3.827182e+05
75%,750.250000,710.250000,7.603211e+05
max,1000.000000,1000.000000,1.972127e+06


## 2. Splitting the dataset

In [58]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df[['Temperature °C', 'Mols KCL']]
y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [76]:
# Use sklearn to train a model on the training set
lin_model = LinearRegression()
lin_model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
sample = pd.DataFrame({
    'Temperature °C': [40],
    'Mols KCL': [0.5]
})
y_pred = lin_model.predict(sample)

print(f"Predicted size {y_pred[0]:.2f} nm^3")
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
train_score = lin_model.score(X_train, y_train)
test_score = lin_model.score(X_test, y_test)

print(f"Training score: {train_score:.4f}")
print(f"Test score: {test_score:.4f}")

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

print(f"intercept: {lin_model.intercept_}")
print(f"coefficients: [Temperature, Mols KCL]: {lin_model.coef_}")


Predicted size -374229.28 nm^3
Training score: 0.8610
Test score: 0.8552
intercept: -409391.4795834075
coefficients: [Temperature, Mols KCL]: [ 866.14641337 1032.69506649]


Write the linear equation of a slime: (example equation: $E = mc^2$)

$$h(x) = -409391.4796 + 866.1464 \times \text{Temperature (°C)} + 1032.6951 \times \text{Mols KCL}$$

Report on score and explain meaning:

For this model, it has training score of 0.8610 and test score of 0.8552, which is the R² value of the model. This implies that about 86% of the variation in the slime size is explained by temperature and potassium chloride concentration. The similarity in training and test scores shows that the model generalizes well to new data without overfitting. With the coefficients is positive shows that temperature and concentration of KCl both positively have an impact and influence on the slime growth and the size.

## 4. Use Cross Validation

In [77]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

k_fold = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(lin_model, X, y, cv=k_fold)

# Report on their finding and their significance
print(f"Cross-validation: {cv_scores}")
print(f"Mean: {cv_scores.mean():.4f}")
print(f"Standard deviation: {cv_scores.std():.4f}")

Cross-validation: [0.61055833 0.75361406 0.29462604 0.6504651  0.68270399]
Mean: 0.5984
Standard deviation: 0.1590


Write findings here:
The cross validation shows that there is consistent level of performance from different splits of data with mean approximately 0.86 and standard deviation of 0.02. This shows that our model is not only stable but also reliable.

## 5. Using Polynomial Regression

In [109]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
X_poly_train, X_poly_test, y_poly_train, y_poly_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)


poly_model = LinearRegression()
poly_model.fit(X_poly_train, y_poly_train)


poly_train_score = poly_model.score(X_poly_train, y_poly_train)
poly_test_score = poly_model.score(X_poly_test, y_poly_test)
print(f"Training score: {poly_train_score:.4f}")
print(f"Test score: {poly_test_score:.4f}")

poly_cv_scores = cross_val_score(poly_model, X_poly, y, cv=k_fold)
print(f"Polynomial CV: {poly_cv_scores}")
print(f"Mean CV: {poly_cv_scores.mean():.4f}")
print(f"Standard deviation: {poly_cv_scores.std():.4f}")

c_poly = poly_model.coef_
i_poly = poly_model.intercept_
        
c_poly.sort()
c_poly = c_poly[::-1]

# Report on the metrics and output the resultant equation as you did in Part 3.
print(f"intercept: {i_poly:.5f}")
formatted_coefficients = ", ".join([f"{coef:.5f}" for coef in c_poly])
print(f"coefficients: [Temperature, Mols KCL]: [{formatted_coefficients}]")

Training score: 0.8334
Test score: 0.8400
Polynomial CV: [0.69121588 0.69407705 0.73980045 0.61215057 0.4412892 ]
Mean CV: 0.6357
Standard deviation: 0.1055
intercept: 1.34671
coefficients: [Temperature, Mols KCL]: [5.23073, 1.59390, 1.33536, 0.89211, 0.75189, 0.52993, 0.46681, 0.41420, 0.24196, 0.21276, 0.12186, 0.10035, 0.10035, 0.07911, 0.00000, 0.00000, -0.02920, -0.10035, -0.10035, -0.22308, -0.42530, -0.61271, -0.68076, -0.83579, -1.34834, -2.10818, -2.21357, -3.44354]


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$$h(x) = 1.34671 + 5.23073 \times \text{Temperature} + 1.59390 \times \text{Mols KCL} + [\text{additional polynomial terms}]$$

Report on the score and interpret:

The polynomial regression model shows good predictive power with a training score of 0.8334 and a test score of 0.8400, which is  lower than the linear regression model. The large variability of cross-validation results (R² mean of 0.6357 with standard deviation of 0.1055) suggests overfitting and inconsistent model performance on different subsets of data. Thus, the linear regression model is more reliable and generalizable for the prediction of slime size based on temperature and KCl concentration.


# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [62]:
# Load the dataset. Then train and evaluate the classification models.
ckd_df = pd.read_csv('ckd_feature_subset.csv')
ckd_df.head(15)

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
0,0.688312,0.333333,0.000000,1,0,0.000000,1
1,0.545455,0.333333,0.128319,1,0,0.305085,1
2,0.714286,0.500000,0.238938,1,0,0.186441,1
3,0.688312,0.333333,0.283186,0,1,0.338983,1
4,0.441558,0.333333,0.221239,1,0,0.220339,1
5,0.740260,0.833333,0.265487,0,1,0.355932,1
6,0.870130,0.333333,0.668142,0,1,0.237288,1
7,0.870130,0.500000,0.150442,0,1,0.372881,1
8,0.194805,0.666667,0.380531,0,1,0.305085,1
9,0.753247,0.833333,0.163717,0,1,0.220339,1


In [63]:
ckd_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153 entries, 0 to 152
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   age         153 non-null    float64
 1   bp          153 non-null    float64
 2   wbcc        153 non-null    float64
 3   appet_poor  153 non-null    int64  
 4   appet_good  153 non-null    int64  
 5   rbcc        153 non-null    float64
 6   Target_ckd  153 non-null    int64  
dtypes: float64(4), int64(3)
memory usage: 8.5 KB


In [64]:
ckd_df.describe()

,age,bp,wbcc,appet_poor,appet_good,rbcc,Target_ckd
count,153.000000,153.000000,153.000000,153.000000,153.000000,153.000000,153.000000
mean,0.563280,0.400871,0.206056,0.124183,0.875817,0.472361,0.274510
std,0.202485,0.183809,0.140267,0.330873,0.330873,0.174521,0.447733
min,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,0.428571,0.166667,0.119469,0.000000,1.000000,0.406780,0.000000
50%,0.571429,0.500000,0.172566,0.000000,1.000000,0.474576,0.000000
75%,0.701299,0.500000,0.261062,0.000000,1.000000,0.593220,1.000000
max,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000


## Results and Conclusion for Classification Experiments

In [92]:
X_ckd = ckd_df.drop('Target_ckd', axis=1)
y_ckd = ckd_df['Target_ckd']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

models = {
    'Logistic Regression': LogisticRegression(random_state=42),
    'Support Vector Machine': SVC(random_state=42),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(max_iter=1000, random_state=42)
}


In [99]:

output = {}

for model_name, i in models.items():
    cv_scores = cross_val_score(i, X_scaled, y, cv=5, scoring='accuracy')
    output[model_name] = {
        'Mean Accuracy': cv_scores.mean(),
        'Standard Deviation': cv_scores.std()
    }
print(pd.DataFrame(output).T)

                        Mean Accuracy  Standard Deviation
Logistic Regression          0.941290            0.024040
Support Vector Machine       0.954194            0.033919
k-Nearest Neighbors          0.941075            0.052810
Neural Network               0.967312            0.021088


Four models were ranked on the basis of their performance to forecast chronic kidney disease. The neural network is the top with 96.73% and a relatively low standard deviation of 0.021. The support vector machine came next with an accuracy of 95.42%. Logistic regression and k-Nearest Neighbors performed equally well, with both of them recording an accuracy of approximately 94%. The application of cross-validation highlighted the consistency of model performance across various data partitions, which indicates good predictive effectiveness.


## Results and Conclusion for Neural Network Experiments

In [111]:
n_layer = {
    '1 layer, 10 neurons': MLPClassifier(hidden_layer_sizes=(10,), max_iter=2000, random_state=42),
    '2 layers, 10 neurons each': MLPClassifier(hidden_layer_sizes=(10, 10), max_iter=1000, random_state=42), 
    '3 layers, 10 neurons each': MLPClassifier(hidden_layer_sizes=(10, 10, 10), max_iter=1000, random_state=42),
    '1 layer, 50 neurons': MLPClassifier(hidden_layer_sizes=(50,), max_iter=2000, random_state=42)
}  
output_n = {}

for config_name, model in n_layer.items():
    cv_scores = cross_val_score(model, X_scaled, y, cv=5, scoring='accuracy')
    output_n[config_name] = {
        'Mean Accuracy': cv_scores.mean(),
        'Standard Deviation': cv_scores.std()
    }

print(pd.DataFrame(output_n).T)



                           Mean Accuracy  Standard Deviation
1 layer, 10 neurons             0.967312            0.021088
2 layers, 10 neurons each       0.973978            0.013017
3 layers, 10 neurons each       0.960860            0.024610
1 layer, 50 neurons             0.973763            0.024835


I experimented with various hidden layer structures and number of neurons with the hope of finding the best model structure. The two-layered network with 10 neurons in each of the two layers performed optimally at the maximum accuracy of 97.40%. A single-layered network with 50 neurons followed closely at 97.38%. The inclusion of a third layer reduced performance by a small amount to 96.09%. This result emphasizes that larger models do not always mean better predictive performance, and there is an optimal level of network complexity for this task of chronic kidney disease prediction.